File: Arvore/arvore.cpp
#include <iostream>
#include <chrono> // Para medir o tempo

using namespace std;

struct Node {
    int key;
    Node* left;
    Node* right;

    // Construtor
    Node(int k) : key(k), left(nullptr), right(nullptr) {}
};

// Função para inserir um novo nó na BST
// Retorna a nova raiz da subárvore
Node* insert(Node* node, int key) {
    // Se a árvore/subárvore estiver vazia, cria um novo nó
    if (node == nullptr) {
        return new Node(key);
    }

    // Caso contrário, desce pela árvore
    if (key < node->key) {
        node->left = insert(node->left, key);
    } else if (key > node->key) {
        node->right = insert(node->right, key);
    }

    // Retorna o ponteiro do nó (inalterado)
    return node;
}

// Função para buscar uma chave na BST
// Retorna true se a chave for encontrada, false caso contrário
bool search(Node* root, int key) {
    Node* current = root;
    while (current != nullptr) {
        if (key == current->key) {
            return true; // Encontrou
        } else if (key < current->key) {
            current = current->left;
        } else {
            current = current->right;
        }
    }
    return false; // Não encontrou
}

// Função para deletar a árvore e liberar memória (evitar memory leaks)
void deleteTree(Node* node) {
    if (node == nullptr) {
        return;
    }
    deleteTree(node->left);
    deleteTree(node->right);
    delete node;
}

int main() {
    Node* root = nullptr;

    //PARÂMETROS PARA AJUSTAR A INTENSIDADE
    
    // Número de nós a inserir na árvore
    // Inserir em ordem cria o pior caso (lista encadeada)
    const int NUM_NODES = 1000000; // 1 milhão de nós

    // Número de buscas a realizar.
    // Buscamos por um valor inexistente para forçar a busca completa
    const long long NUM_SEARCHES = 500000000; // 500 milhões de buscas
    

    cout << "Iniciando criacao da arvore desbalanceada com " << NUM_NODES << " nos..." << endl;
    
    auto start_insert = chrono::high_resolution_clock::now();
    for (int i = 0; i < NUM_NODES; ++i) {
        root = insert(root, i);
    }
    auto end_insert = chrono::high_resolution_clock::now();
    chrono::duration<double> insert_duration = end_insert - start_insert;
    cout << "Arvore criada em " << insert_duration.count() << " segundos." << endl;


    cout << "Iniciando " << NUM_SEARCHES << " buscas (pior caso)..." << endl;
    cout << "(Isso pode demorar e vai usar bastante CPU)" << endl;

    auto start_search = chrono::high_resolution_clock::now();
    
    volatile bool result = false; // volatile para evitar otimizações do compilador
    
    for (long long i = 0; i < NUM_SEARCHES; ++i) {
        // Buscamos por -1, que não existe, forçando a percorrer 
        // o caminho mais longo (até o fim da "lista") todas as vezes.
        result = search(root, -1); 
    }
    
    auto end_search = chrono::high_resolution_clock::now();
    chrono::duration<double> search_duration = end_search - start_search;
    
    cout << "Buscas concluidas." << endl;
    cout << "Tempo total das buscas: " << search_duration.count() << " segundos." << endl;

    // Limpa a memória
    cout << "Limpando memoria..." << endl;
    deleteTree(root);
    cout << "Concluido." << endl;

    return 0;
}

File: Arvore/arvore.py
import time
import sys

# Definição do nó da árvore
class Node:
    def __init__(self, key):
        self.key = key
        self.left = None
        self.right = None

# Função para inserir um novo nó na BST (de forma iterativa)
# Isso evita o estouro do limite de recursão do Python
def insert(root, key):
    new_node = Node(key)
    
    # Se a árvore está vazia, o novo nó é a raiz
    if root is None:
        return new_node
    
    current = root
    while True:
        if key < current.key:
            # Vai para a esquerda
            if current.left is None:
                current.left = new_node
                return root # Retorna a raiz original
            current = current.left
        else:
            # Vai para a direita (ou igual)
            if current.right is None:
                current.right = new_node
                return root # Retorna a raiz original
            current = current.right

# Função para buscar uma chave na BST (de forma iterativa)
def search(root, key):
    current = root
    while current is not None:
        if key == current.key:
            return True # Encontrou
        elif key < current.key:
            current = current.left
        else:
            current = current.right
    return False # Não encontrou

# --- PARÂMETROS PARA AJUSTAR A INTENSIDADE ---

# AVISO: Estes números são BEM MENORES que a versão C++.
# Python é muito mais lento para esta tarefa.

# O custo de construção é O(N^2), então 50.000 já é pesado.
NUM_NODES = 50000  # 50 mil nós (em vez de 1 milhão)

# O custo de busca é O(M*N).
NUM_SEARCHES = 100000 # 100 mil buscas (em vez de 500 milhões)

# ---------------------------------------------

# Python não tem um limite de "int" como C, mas 
# é bom saber o limite de recursão (que evitamos)
# print(f"Limite de recursao do Python: {sys.getrecursionlimit()}")

root = None

print(f"Iniciando criacao da arvore desbalanceada com {NUM_NODES} nos...")
print("(Isso ja vai ser O(N^2) e pode demorar um pouco)...")

start_insert = time.perf_counter() # Medidor de tempo mais preciso

# Inserindo em ordem para forçar o pior caso (O(N^2) no total)
for i in range(NUM_NODES):
    root = insert(root, i)
    
end_insert = time.perf_counter()
insert_duration = end_insert - start_insert
print(f"Arvore criada em {insert_duration:.4f} segundos.")


print(f"Iniciando {NUM_SEARCHES} buscas (pior caso)...")
print("(Isso pode demorar e vai usar bastante CPU)")

start_search = time.perf_counter()

result = False
for i in range(NUM_SEARCHES):
    # Buscamos por -1, que não existe, forçando a percorrer 
    # o caminho mais longo (O(N)) todas as vezes.
    result = search(root, -1) 

end_search = time.perf_counter()
search_duration = end_search - start_search

print("Buscas concluidas.")
print(f"Tempo total das buscas: {search_duration:.4f} segundos.")

# Em Python, a limpeza de memória (Garbage Collection) é automática.
# Apenas definimos root como None para sinalizar que não usamos mais.
root = None 
print("Concluido.")